### PR TITLE
CL: add swap logic

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

Adds SwapExactAmountIn/Out are respective logic for bank sends to the concentrated liquidity module

This likely will need one or two more iterations, but wanted to get some reviews on this so we can merge to main and get rid of the concentrated-liquidity-main branch.

### CODE REVIEW

File: x/gamm/types/pool.go [hunk 1]
```diff
@@ -141,6 +137,16 @@ type WeightedPoolExtension interface {
...
+type BalancerPool interface {
+	PoolI
+	TraditionalAmmInterface
+}
+
+type StableSwapExtension interface {
+	PoolI
+	TraditionalAmmInterface
+}
+
```
Comment (czarcas7ic) [inline x/gamm/types/pool.go:149]: Note to reviewer, need to decide what we are adding to these extensions. Was not clear to me what is not shared since I did not work on stable swap

File: x/concentrated-liquidity/types/pool.go [hunk 2]
```diff
@@ -3,10 +3,12 @@ package types
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	proto "github.com/gogo/protobuf/proto"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v13/x/gamm/types"
 )
 
 type ConcentratedPoolExtension interface {
-	// gammtypes.PoolI
+	gammtypes.PoolI
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/pool.go:11]: We should remove the `PoolI` methods redefined below

File: x/swaprouter/router.go [hunk 3]
```diff
@@ -130,7 +132,12 @@ func (k Keeper) MultihopEstimateOutGivenExactAmountIn(
 			swapFee = routeSwapFee.Mul((swapFee.Quo(sumOfSwapFees)))
 		}
 
-		tokenOut, err := pool.CalcOutAmtGivenIn(ctx, sdk.Coins{tokenIn}, route.TokenOutDenom, swapFee)
+		poolTrad, ok := pool.(gammtypes.TraditionalAmmInterface)
```
Comment (p0mvn) [inline x/swaprouter/router.go:135]: Is it possible to make this more general to work with CL pool as well?

We should aim to have everything in `swaprouter` to be generalized for any kind of pool

File: x/gamm/types/pool.go (hunk 1 above)
Comment (p0mvn) [reply x/gamm/types/pool.go:149]: Most logic is shared and we shouldn't need these.

Balancer pool supports more methods but those are defined on other / additional interface extensions such as `PoolAmountOutExtension`

File: x/gamm/keeper/swap.go [hunk 4]
```diff
@@ -31,6 +31,11 @@ func (k Keeper) SwapExactAmountIn(
 	}
 	tokensIn := sdk.Coins{tokenIn}
 
+	pool, ok := poolI.(types.TraditionalAmmInterface)
+	if !ok {
+		return sdk.Int{}, errors.New("type cast failed")
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:36]: consider formatting actual and expected types and also covering this with a test

File: x/gamm/keeper/pool.go [hunk 5]
```diff
@@ -120,14 +125,20 @@ func (k Keeper) GetPoolsAndPoke(ctx sdk.Context) (res []types.TraditionalAmmInte
...
 		}
 
-		if pokePool, ok := pool.(types.WeightedPoolExtension); ok {
+		if pokePool, ok := poolI.(types.WeightedPoolExtension); ok {
 			pokePool.PokePool(ctx.BlockTime())
 		}
+
+		pool, ok := poolI.(types.TraditionalAmmInterface)
+		if !ok {
+			return nil, fmt.Errorf("type cast failed")
```
Comment (p0mvn) [inline x/gamm/keeper/pool.go:139]: ditto for error. Maybe we could define this error in x/gamm/types/errors.go and reuse where applicable?

File: x/concentrated-liquidity/types/expected_keepers.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+// creating a x/concentrated-liquidity keeper.
+type BankKeeper interface {
+	SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) error
+
+	SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) error
+
+	SendCoins(ctx sdk.Context, fromAddr sdk.AccAddress, toAddr sdk.AccAddress, amt sdk.Coins) error
+
+	MintCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	BurnCoins(ctx sdk.Context, name string, amt sdk.Coins) error
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/expected_keepers.go:18]: Do we need these?

File: x/gamm/keeper/swap.go (hunk 4 above)
Comment (czarcas7ic) [reply x/gamm/keeper/swap.go:36]: Im not quite sure how to go about formatting the error to show an "expected" vs "actual" type without just hard coding it.

File: x/gamm/keeper/pool.go (hunk 5 above)
Comment (czarcas7ic) [reply x/gamm/keeper/pool.go:139]: Same question as above

File: x/concentrated-liquidity/types/expected_keepers.go (hunk 6 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/types/expected_keepers.go:18]: Removed

File: x/swaprouter/router.go (hunk 3 above)
Comment (czarcas7ic) [reply x/swaprouter/router.go:135]: Can we talk about this further? Do you for see just checking if the msg is CL and use the CL keeper method, otherwise use the TraditionalAmmInterface CalcOut?

File: x/gamm/keeper/swap.go (hunk 4 above)
Comment (p0mvn) [reply x/gamm/keeper/swap.go:36]: I'd do something like

```
 fmt.Errorf("failed cast to TraditionalAmmInterface, actual type: %T, poolI)
```

Comment (p0mvn) [reply x/gamm/keeper/swap.go:36]: but not a blocker, feel free to skip if you want

File: x/swaprouter/router.go (hunk 3 above)
Comment (p0mvn) [reply x/swaprouter/router.go:135]: I would instead propagate this logic to the keeper (`swapModule`). Similar to how we route swaps. The keeper methods can deal with that instead

File: x/gamm/keeper/swap.go (hunk 4 above)
Comment (czarcas7ic) [reply x/gamm/keeper/swap.go:36]: ah, duh, thanks 😅

File: x/concentrated-liquidity/lp.go [hunk 7]
```diff
@@ -91,6 +106,26 @@ func (k Keeper) withdrawPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAd
...
+	// send withdraw amount from pool to position owner
+	if actualAmount0.IsPositive() && actualAmount1.IsPositive() {
+		err = k.bankKeeper.SendCoins(ctx, pool.GetAddress(), owner, sdk.NewCoins(sdk.NewCoin(denom0, actualAmount0), sdk.NewCoin(denom1, actualAmount1)))
+	} else if actualAmount0.IsPositive() && actualAmount1.IsZero() {
+		err = k.bankKeeper.SendCoins(ctx, pool.GetAddress(), owner, sdk.NewCoins(sdk.NewCoin(denom0, actualAmount0)))
+	} else if actualAmount0.IsZero() && actualAmount1.IsPositive() {
+		err = k.bankKeeper.SendCoins(ctx, pool.GetAddress(), owner, sdk.NewCoins(sdk.NewCoin(denom1, actualAmount0)))
+	}
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, err
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:126]: Note to reviewer: does a normal send coins work here or do we need to eventually change this to a module account send

File: x/concentrated-liquidity/lp.go [hunk 8]
```diff
@@ -56,6 +56,21 @@ func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
 		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, types.InsufficientLiquidityCreatedError{Actual: actualAmount1, Minimum: amount1Min}
 	}
 
+	denom0 := pool.GetToken0()
+	denom1 := pool.GetToken1()
+
+	// send deposit amount from position owner to pool
+	if actualAmount0.IsPositive() && actualAmount1.IsPositive() {
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:63]: Is this a requirement to be implemented for testing swaps? If not, can we submit this in another PR to keep this one focused on swaps only and to avoid blocking it on this?

Some notes:
- I'm not sure why we have so many if /else. Should be possible to:

```
if amount 0 is not positive => error
if amount 1 is not positive => error

bank send denom0 with amount 0
bank send denom 1 with amount 1
```
- this logic should be tested, might also make sense to encapsulate it in a helper

File: x/concentrated-liquidity/lp.go (hunk 7 above)
Comment (p0mvn) [reply x/concentrated-liquidity/lp.go:126]: `SendCoinsFromModuleToAccount`

File: x/concentrated-liquidity/internal/emit.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v13/x/gamm/types"
+)
+
+// TODO: change as needed for CL rather than gamm
+func EmitSwapEvent(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, input sdk.Coin, output sdk.Coin) {
```
Comment (p0mvn) [inline x/concentrated-liquidity/internal/emit.go:12]: Can we instead move the swap event logic to swap router, make it public and import it to gamm and here?

File: x/concentrated-liquidity/internal/emit.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+func newRemoveLiquidityEvent(sender sdk.AccAddress, poolId uint64, liquidity sdk.Coins) sdk.Event {
+	return sdk.NewEvent(
+		types.TypeEvtPoolExited,
+		sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+		sdk.NewAttribute(sdk.AttributeKeySender, sender.String()),
+		sdk.NewAttribute(types.AttributeKeyPoolId, strconv.FormatUint(poolId, 10)),
+		sdk.NewAttribute(types.AttributeKeyTokensOut, liquidity.String()),
+	)
+}
```
Comment (p0mvn) [inline x/concentrated-liquidity/internal/emit.go:71]: I don't think we need these yet. Only swaps should be needed for now to make the tests pass. We should define concentrated-liqudity specific events separately

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 11]
```diff
@@ -857,7 +857,10 @@ func (suite *KeeperTestSuite) TestJoinPool() {
 		tc := tc
 
 		suite.T().Run(tc.name, func(t *testing.T) {
-			pool := createTestPool(t, tc.swapFee, sdk.ZeroDec(), tc.poolAssets...)
+			poolI := createTestPool(t, tc.swapFee, sdk.ZeroDec(), tc.poolAssets...)
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/pool_suite_test.go:860]: let's instead make the `createTestPool` return balancer right away instead of type casting everywhere

File: x/concentrated-liquidity/types/expected_keepers.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+// BankKeeper defines the banking contract that must be fulfilled when
+// creating a x/concentrated-liquidity keeper.
+type BankKeeper interface {
+	SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) error
+
+	SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) error
+
+	SendCoins(ctx sdk.Context, fromAddr sdk.AccAddress, toAddr sdk.AccAddress, amt sdk.Coins) error
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/expected_keepers.go:15]: Do we need all 3?

File: x/concentrated-liquidity/lp.go (hunk 7 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go:126]: I actually dont think we use FromModuleToAccount here. I believe in gamm we use SendCoins. Plus SendCoinsFromModuleToAccount only takes a module name, which we wouldn't be able to do for each pool

File: x/concentrated-liquidity/internal/model/pool.go [hunk 13]
```diff
@@ -182,3 +182,25 @@ func (p *Pool) ApplySwap(newLiquidity sdk.Dec, newCurrentTick sdk.Int, newCurren
...
+
+// TODO: Figure out what we are going to do with these stubs since we call these in the keeper but are needed to use PoolI
+func (p Pool) CalcInAmtGivenOut(
+	ctx sdk.Context, tokensOut sdk.Coins, tokenInDenom string, swapFee sdk.Dec) (tokenIn sdk.Coin, err error) {
+	return sdk.Coin{}, nil
+}
+
+func (p Pool) CalcOutAmtGivenIn(ctx sdk.Context, tokensIn sdk.Coins, tokenOutDenom string, swapFee sdk.Dec) (sdk.Coin, error) {
+	return sdk.Coin{}, nil
+}
```
Comment (p0mvn) [inline x/concentrated-liquidity/internal/model/pool.go:194]: Why? We shouldn't need these as they should be defined on the `TraditionalAmmInterface`

File: x/concentrated-liquidity/internal/model/pool.go [hunk 14]
```diff
@@ -182,3 +182,25 @@ func (p *Pool) ApplySwap(newLiquidity sdk.Dec, newCurrentTick sdk.Int, newCurren
...
+	return sdk.Coins{}
+}
+
+func (p *Pool) SwapInAmtGivenOut(ctx sdk.Context, tokensOut sdk.Coins, tokenInDenom string, swapFee sdk.Dec) (tokenIn sdk.Coin, err error) {
+	return sdk.Coin{}, nil
+}
+
+func (p *Pool) SwapOutAmtGivenIn(ctx sdk.Context, tokensIn sdk.Coins, tokenOutDenom string, swapFee sdk.Dec) (tokenOut sdk.Coin, err error) {
+	return sdk.Coin{}, nil
+}
```
Comment (p0mvn) [inline x/concentrated-liquidity/internal/model/pool.go:206]: These too

File: x/concentrated-liquidity/swaps.go [hunk 15]
```diff
@@ -47,13 +52,53 @@ func (k Keeper) CreateNewConcentratedLiquidityPool(ctx sdk.Context, poolId uint6
...
 	tokenOutMinAmount sdk.Int,
 	swapFee sdk.Dec,
-) (sdk.Int, error) {
-	return sdk.Int{}, nil
+) (tokenOutAmount sdk.Int, err error) {
+	if tokenIn.Denom == tokenOutDenom {
+		return sdk.Int{}, errors.New("cannot trade same denomination in and out")
+	}
+
+	// get concentratedPool extension from poolI
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:65]: Let's do a type cast instead of accessing the store

File: x/concentrated-liquidity/swaps.go [hunk 16]
```diff
@@ -47,13 +52,53 @@ func (k Keeper) CreateNewConcentratedLiquidityPool(ctx sdk.Context, poolId uint6
...
+	asset0 := pool.GetToken0()
+	zeroForOne := tokenIn.Denom == asset0
+	var tokenOutCoin sdk.Coin
+
+	// change priceLimit based on which direction we are swapping
+	if zeroForOne {
+		tokenOutCoin, err = k.SwapOutAmtGivenIn(ctx, tokenIn, tokenOutDenom, swapFee, sdk.NewDec(1), pool.GetId())
+	} else {
+		tokenOutCoin, err = k.SwapOutAmtGivenIn(ctx, tokenIn, tokenOutDenom, swapFee, sdk.NewDec(999999999999), pool.GetId())
+	}
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:81]: Let's define `sdk.NewDec(999999999999)` as `upperPriceLimit` and `sdk.OneDec()` as `lowerPriceLimit` 

```suggestion
        var priceLimit sdk.Dec
	// change priceLimit based on which direction we are swapping
	if zeroForOne {
		priceLimit = lowerPriceLimit
	} else {
		priceLimit = upperPriceLimit
	}
	tokenOutCoin, err = k.SwapOutAmtGivenIn(ctx, tokenIn, tokenOutDenom, swapFee, priceLimit, pool.GetId())
```

File: x/concentrated-liquidity/swaps.go [hunk 17]
```diff
@@ -47,13 +52,53 @@ func (k Keeper) CreateNewConcentratedLiquidityPool(ctx sdk.Context, poolId uint6
...
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// check that the tokenOut calculated is both valid and less than specified limit
+	tokenOutAmount = tokenOutCoin.Amount
+	if !tokenOutAmount.IsPositive() {
+		return sdk.Int{}, sdkerrors.Wrapf(gammtypes.ErrInvalidMathApprox, "token amount must be positive")
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:89]: Let's avoid using sdkerrors please

File: x/concentrated-liquidity/swaps.go [hunk 18]
```diff
@@ -84,10 +129,31 @@ func (k Keeper) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDen
...
+	swapFee sdk.Dec,
+) (tokenIn sdk.Coin, err error) {
+	return sdk.Coin{}, nil
+}
+
+// CalcOutAmtGivenInOld calculates tokens to be swapped out given the provided amount and fee deducted. It also returns
 // what the updated tick, liquidity, and currentSqrtPrice for the pool would be after this swap.
 // Note this method is non-mutative, so the values returned by CalcOutAmtGivenIn do not get stored
-func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
+func (k Keeper) CalcOutAmtGivenInOld(ctx sdk.Context,
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:156]: Let's call this `calcOutAmtGivenIn`.

File: x/concentrated-liquidity/swaps.go [hunk 19]
```diff
@@ -242,7 +308,7 @@ func (k *Keeper) SwapInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coin, tokenInDe
 	return tokenInCoin, nil
 }
 
-func (k Keeper) CalcInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coin, tokenInDenom string, swapFee sdk.Dec, minPrice, maxPrice sdk.Dec, poolId uint64) (sdk.Coin, sdk.Dec, sdk.Int, sdk.Dec, error) {
+func (k Keeper) CalcInAmtGivenOutOld(ctx sdk.Context, tokenOut sdk.Coin, tokenInDenom string, swapFee sdk.Dec, minPrice, maxPrice sdk.Dec, poolId uint64) (sdk.Coin, sdk.Dec, sdk.Int, sdk.Dec, error) {
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:311]: ditto for lowercase

File: x/gamm/keeper/swap.go [hunk 20]
```diff
@@ -123,6 +123,36 @@ func (k Keeper) SwapExactAmountOut(
 	return tokenInAmount, nil
 }
 
+func (k Keeper) CalcOutAmtGivenIn(
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:126]: Let's add basic success and failure test cases. To cover given pool not implementing `TraditionalAmmInterface`

File: x/swaprouter/router.go [hunk 21]
```diff
@@ -377,13 +379,18 @@ func (k Keeper) createOsmoMultihopExpectedSwapOuts(
...
 		if err != nil {
 			return nil, err
 		}
 
-		swapFee := pool.GetSwapFee(ctx)
-		tokenIn, err := pool.CalcInAmtGivenOut(ctx, sdk.NewCoins(tokenOut), route.TokenInDenom, cumulativeRouteSwapFee.Mul((swapFee.Quo(sumOfSwapFees))))
+		poolTrad, ok := poolI.(gammtypes.TraditionalAmmInterface)
+		if !ok {
+			return nil, fmt.Errorf("failed cast to TraditionalAmmInterface, actual type: %T", poolI)
+		}
```
Comment (p0mvn) [inline x/swaprouter/router.go:390]: Should this cast be removed?

File: x/concentrated-liquidity/internal/model/pool.go (hunk 13 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/model/pool.go:194]: They are defined in PoolI within swaprouter here, not TraditionalAmmInterface

https://github.com/osmosis-labs/osmosis/blob/dba9df0db5a8330de4204e52ec0369cea31c570c/x/swaprouter/types/pool.go#L30-L42

Comment (p0mvn) [reply x/concentrated-liquidity/internal/model/pool.go:194]: Ahh I see, these should be defined on the `TraditionalAmmInterface`. I'll try to see if moving these is a quick change. If so, maybe we can merge that change first to avoid creating the stubs here unnecessarily

Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/model/pool.go:194]: Didn't we have these defined on TAMM previously and then decided to move them up to PoolI?

Comment (p0mvn) [reply x/concentrated-liquidity/internal/model/pool.go:194]: No, based on the component diagram, the goal has been to have it on `TraditionalAmmInterface`. The only reason why we would keep these methods on `PoolI` until now it is to be able to port `EstimateSwap` to swaprouter. However, since we now call keeper's "Calc" method instead of the pool's, we can move these to the right place.

I've looked into this problem, and moving the methods to `TraditionalAmmInterface` depends on this PR. Specificcally, on the change I mentioned above (where swaprouter uses keeper's calc as opposed to pool's).

So I think we can merge this first and refactor separately.

File: x/concentrated-liquidity/swaps_test.go [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+
+	tests := []struct {
+		name      string
+		param     param
+		expectErr bool
+	}{
+		{
+			name: "Proper swap foo > bar",
+			param: param{
+				tokenIn:           sdk.NewCoin("foo", sdk.NewInt(42000000)),
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:44]: Is there a Wolfram link you are using to QA this? How can I be sure that the values are correct here?

File: x/concentrated-liquidity/swaps_test.go [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+				tokenIn:           sdk.NewCoin("bar", sdk.NewInt(13370)),
+				tokenOutDenom:     "foo",
+				tokenOutMinAmount: sdk.NewInt(1),
+				expectedTokenOut:  sdk.NewInt(66808387),
+			},
+		},
+		{
+			name: "out is lesser than min amount",
+			param: param{
+				tokenIn:           sdk.NewCoin("foo", sdk.NewInt(42000000)),
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:62]: nit: seems that a  lot of the same values are reused throughout the test - `42000000`, `8397` etc. Consider initializing constants with these values and reusing them where applicable

File: x/concentrated-liquidity/swaps_test.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+			s.Require().NoError(err)
+
+			// Check the test case to see if we are swapping asset0 for asset1 or vice versa
+			asset0 := pool.GetToken0()
+			zeroForOne := test.param.tokenIn.Denom == asset0
+
+			// Fund the test account with foo and bar, then create a default position to the pool created earlier
+			s.FundAcc(s.TestAccs[0], sdk.NewCoins(sdk.NewCoin("bar", sdk.NewInt(10000000000000)), sdk.NewCoin("foo", sdk.NewInt(1000000000000))))
+			_, _, _, err = s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, 1, s.TestAccs[0], defaultAmt0, defaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), lowerTick.Int64(), upperTick.Int64())
+			s.Require().NoError(err)
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:113]: If this is a default positon and assuming that this is going to be reused frequently, should we make a test helper for this?

File: x/concentrated-liquidity/swaps_test.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+
+			if test.expectErr {
+				pool, err = s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, pool.GetId())
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:116]: Why do we get pool here again? Seems that we have one initialized above and the "else" case does not re-get it

File: x/concentrated-liquidity/swaps_test.go [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+				prevGasConsumed := s.Ctx.GasMeter().GasConsumed()
+
+				// Execute the swap directed in the test case
+				tokenOutAmount, err := s.App.ConcentratedLiquidityKeeper.SwapExactAmountIn(s.Ctx, s.TestAccs[0], pool.(swaproutertypes.PoolI), test.param.tokenIn, test.param.tokenOutDenom, test.param.tokenOutMinAmount, swapFee)
+				s.Require().NoError(err)
+				s.Require().Equal(test.param.expectedTokenOut.String(), tokenOutAmount.String())
+
+				gasConsumedForSwap := s.Ctx.GasMeter().GasConsumed() - prevGasConsumed
+				// TODO: make a CLGasFeeForSwap
+				// Check that we consume enough gas that a CL pool swap warrants
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:130]: nit: seems like a quick change. Do you mind adding it right away instead of having to track this?

File: x/concentrated-liquidity/swaps_test.go [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+
+				spotPriceAfter := pool.GetCurrentSqrtPrice().Power(2)
+
+				// Ratio of the token out should be between the before spot price and after spot price.
+				var tradeAvgPrice sdk.Dec
+				if zeroForOne {
+					tradeAvgPrice = tokenOutAmount.ToDec().Quo(test.param.tokenIn.Amount.ToDec())
+				} else {
+					tradeAvgPrice = test.param.tokenIn.Amount.ToDec().Quo(tokenOutAmount.ToDec())
+				}
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:149]:
```suggestion
				tradeAvgPrice := tradeAvgPrice = tokenOutAmount.ToDec().Quo(test.param.tokenIn.Amount.ToDec())
				if !zeroForOne {
					tradeAvgPrice = sdk.OneDec().Quo(tradeAvgPrice)
				}
```

**Review by p0mvn — APPROVED** [review]
LGTM once remaining comments are addressed.

File: x/concentrated-liquidity/swaps_test.go (hunk 22 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/swaps_test.go:44]: "trust me bro"

jk, I am using the same base test cases from pool_test, I can copy over the wolfram links from there though

File: x/concentrated-liquidity/swaps_test.go (hunk 23 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/swaps_test.go:62]: I kind of feel displaying the number itself, despite being repeated, is the best way. If it was instead something like "baseCaseEthIn" and "baseCaseUSDCOut", we would just need to look at what that is anyway to confirm the math is right, right? Please lmk if I am misunderstanding, I very well could be

Comment (p0mvn) [reply x/concentrated-liquidity/swaps_test.go:73]: Fair, that makes sense! Let's keep it

File: x/concentrated-liquidity/swaps_test.go (hunk 22 above)
Comment (p0mvn) [reply x/concentrated-liquidity/swaps_test.go:49]: Yeah I think duplicating a link would be helpful. While I assumed that there is probably another test case using the same values, it is just much more easier having it available in this test directly

File: x/gamm/keeper/swap.go (hunk 20 above)
Comment (czarcas7ic) [reply x/gamm/keeper/swap.go:126]: This is the last point I have not addressed, are you saying instead of returning a blank coin and error, you want me to check if we can type cast the provided poolI to TAMM, and then make a test in swaps_test to test for this?

Comment (p0mvn) [reply x/gamm/keeper/swap.go:126]: I just want to make sure that we have test coverage for the following:
- balancer pool given -> swap is done (minimal assertions to ensure correctness)
- stableswap pool given -> swap is done (minimal assertions to ensure correctness)
- cl pool given -> swap not done and error returned

Comment (czarcas7ic) [reply x/gamm/keeper/swap.go:126]: But this comment is point to Calc, not swap. Was this meant to be for Swap?

Comment (p0mvn) [reply x/gamm/keeper/swap.go:126]: On a related note, in another PR, I would like to suggest making a helper:

```go
func convertToCFMMPooll() (types.TraditionalAmmInterface, error) {
    	cfmmPool, ok := poolI.(types.TraditionalAmmInterface)
	if !ok {
		return nil, fmt.Errorf("given pool does not implement TraditionalAmmInterface, implements %T", poolI)
	}
       return cfmmPool, nil
}
```

and a similar helper for the concentrated pool conversion

Comment (p0mvn) [reply x/gamm/keeper/swap.go:126]: @czarcas7ic sorry just saw your reply. I meant calc, not swap. Edited my previous comment

**Review by p0mvn — APPROVED** [review]
